paddle/gserver/layers/MultinomialSampler.cpp

```diff
@@ -49,26 +50,19 @@ MultinomialSampler::MultinomialSampler(const real* prob, int size)
   int smallPos = nextSmallPos(0);
   int bigPos = nextBigPos(0);
 
-  auto fillIntervals = [&]() {
-    while (bigPos < size && smallPos < size) {
-      while (intervals_[bigPos].thresh > 1 && smallPos < size) {
-        intervals_[smallPos].otherId = bigPos;
-        intervals_[bigPos].thresh -= 1 - intervals_[smallPos].thresh;
-        smallPos = nextSmallPos(smallPos + 1);
-      }
-      bigPos = nextBigPos(bigPos + 1);
-      // If intervals_[bigPos].thresh < 1, it becomes a small interval
+  while (smallPos < size && bigPos < size) {
+    while (intervals_[bigPos].thresh >= 1 && smallPos < size) {
+      intervals_[smallPos].otherId = bigPos;
+      intervals_[bigPos].thresh -= 1 - intervals_[smallPos].thresh;
+      smallPos = nextSmallPos(smallPos + 1);
     }
-  };
-
-  fillIntervals();
-
-  smallPos = nextSmallPos(0);
 
-  // At this point there is no small intervals after bigPos. And this condition
-  // will remain true during the next fillIntervals()
-
-  fillIntervals();
+    if (intervals_[bigPos].thresh < 1) {
+      bigPos = nextBigPos(0);
```

Comment:
This will make the complexity O(n^2). The original complexity is O(n) every step either bigPos or smallPos will increase.
Should be:
if (intervals_[bigPos].thresh <=1 ) {
  bigPos = nextBigPos(bigPos + 1);
}